examples/name_kernel.cpp

```diff
@@ -0,0 +1,178 @@
+#if defined(RAJA_DEVICE_ACTIVE)
+                                        ,device_loop
+#endif
+                                        >;
+
+#if defined(RAJA_ENABLE_CUDA)
+using reduce_policy = RAJA::cuda_reduce;
+#elif defined(RAJA_ENABLE_HIP)
+using reduce_policy = RAJA::hip_reduce;
+#elif defined(RAJA_ENABLE_OPENMP)
+using reduce_policy = RAJA::omp_reduce;
+#else
+using reduce_policy = RAJA::seq_reduce;
+#endif
+
+
+int main(int argc, char *argv[])
+{
+
+  if(argc != 2) {
+    RAJA_ABORT_OR_THROW("Usage ./teams_reductions host or ./tut_reductions device");
+  }
+
+  //
+  // Run time policy section is demonstrated in this example by specifying
+  // kernel exection space as a command line argument (host or device).
+  // Example usage ./teams_reductions host or ./teams_reductions device
+  //
+  std::string exec_space = argv[1];
+  if(!(exec_space.compare("host") == 0 || exec_space.compare("device") == 0 )){
+    RAJA_ABORT_OR_THROW("Usage ./teams_reductions host or ./teams_reductions device");
+    return 0;
+  }
+
+  RAJA::expt::ExecPlace select_cpu_or_gpu;
+  if(exec_space.compare("host") == 0)
+    { select_cpu_or_gpu = RAJA::expt::HOST; printf("Running RAJA-Teams reductions example on the host \n"); }
+  if(exec_space.compare("device") == 0)
+    { select_cpu_or_gpu = RAJA::expt::DEVICE; printf("Running RAJA-Teams reductions example on the device \n"); }
+
+  // _reductions_array_init_start
+//
+// Define array length
+//
+  const int N = 1000000;
+
+//
+// Allocate array data and initialize data to alternating sequence of 1, -1.
+//
+  int* a = memoryManager::allocate<int>(N);
+
+  for (int i = 0; i < N; ++i) {
+    if ( i % 2 == 0 ) {
+      a[i] = 1;
+    } else {
+      a[i] = -1;
+    }
+  }
+
+//
+// Set min and max loc values
+//
+  const int minloc_ref = N / 2;
+  a[minloc_ref] = -100;
+
+  const int maxloc_ref = N / 2 + 1;
+  a[maxloc_ref] = 100;
+  // _reductions_array_init_end
+
+//
+// Note: with this data initialization scheme, the following results will
+//       be observed for all reduction kernels below:
+//
+//  - the sum will be zero
+//  - the min will be -100
+//  - the max will be 100
+//  - the min loc will be N/2
+//  - the max loc will be N/2 + 1
+//
+//
+
+//
+// Define index range for iterating over a elements in all examples
+//
+  // _reductions_range_start
+  RAJA::RangeSegment arange(0, N);
+  // _reductions_range_end
+
+//----------------------------------------------------------------------------//
+
+  RAJA::ReduceSum<reduce_policy, int> kernel_sum(0);
+  RAJA::ReduceMin<reduce_policy, int> kernel_min(std::numeric_limits<int>::max());
+  RAJA::ReduceMax<reduce_policy, int> kernel_max(std::numeric_limits<int>::min());
+  RAJA::ReduceMinLoc<reduce_policy, int> kernel_minloc(std::numeric_limits<int>::max(), -1);
+  RAJA::ReduceMaxLoc<reduce_policy, int> kernel_maxloc(std::numeric_limits<int>::min(), -1);
+
+  RAJA::forall(RAJA::ExecContext<RAJA::cuda_exec<256>>("My kernel"), arange, [=] __device__ (int i) {
```

Comment:
In the future I'm imagining ExecContext could take arguments like the resource as well as the name. This would let us simplify our API to just have policy by template and ExecContext versions. Then we can handle optional arguments like the name and resource in one place.

Reply:
Be careful with this. Our execution policies sprawled a lot, and got expensive to compile. One thing I recommend is that any "trait" you want in here, make it explicitly tagged for what it is:

```c++
RAJA::ExecContext<RAJA::cuda_exec<256>>(RAJA::KernelLabel("My kernel"), RAJA::SomeOtherTrait(RAJA::gofast))
```

Otherwise you wind up writing really complicated deduction stuff